### ISSUE TITLE
Will imputer be part of the exported pipeline?

### ISSUE THREAD

**Issue by ajing** [rhiever/tpot#511]
My data has some missing values. After running TPOT, the generated pipeline doesn't have any imputer. The model in the final pipeline is ExtraTreeRegressor, which requires no missing values. Is that normal?

**Comment by weixuanfu2016** [comment]
Hmm, the default imputer `Imputer(strategy="median", axis=1)` in TPOT is very simple as a step of preprocessing if imputer is not specified. The current version of TPOT won't export the imputation step. Maybe we should add this step in exported codes as demo below? @rhiever 

```
from sklearn.preprocessing import Imputer
imputer=Imputer(strategy="median", axis=1)
imputer.fit(features)
imputed_features = imputer.transform(features)
```

**Comment by rhiever** [comment]
Seems reasonable to add the imputation step to the exported pipeline if missing data was in the training data. Perhaps we can store an internal boolean variable indicating whether missing data was in the training data, and add the Imputer to the exported pipeline if True?

cc @teaearlgraycold

**Comment by teaearlgraycold** [comment]
Sure, that's easy enough.

**Comment by weixuanfu** [comment]
This new function was already merged to TPOT 0.9, so I closed this issue.

**Comment by vruusmann** [comment]
I'm experimenting with TPOT 0.9.1, and contrary to what is claimed in the above comment, the `fitted_pipeline_` doesn't seem to contain any imputer instances.

My training dataset contains missing values, so during training TPOT reports the following:
```
Imputing missing values in feature set
```

However, when I do `print(repr(regressor.fitted_pipeline_))`, then there's no `Imputer` step anywhere to be seen. Also, the fitted pipeline fails to score a testing dataset, which contains missing values:
```
Best pipeline: DecisionTreeRegressor(RidgeCV(RobustScaler(input_matrix)), max_depth=5, min_samples_leaf=12, min_samples_split=7)
Pipeline(memory=None,
     steps=[('robustscaler', RobustScaler(copy=True, quantile_range=(25.0, 75.0), with_centering=True,
       with_scaling=True)), ('stackingestimator', StackingEstimator(estimator=RidgeCV(alphas=(0.1, 1.0, 10.0), cv=None, fit_intercept=True, gcv_mode=None,
    normalize=False, scoring=None, store_cv_val...lit=7, min_weight_fraction_leaf=0.0,
           presort=False, random_state=None, splitter='best'))])
```

**Comment by rhiever** [comment]
TPOT's imputation step isn't stored in the internal pipelines, but it should apply imputation when you call any functions like fit, predict, score, etc. Also when you export the pipeline, it should export the pipeline with the Imputer at the beginning.

**Comment by vruusmann** [comment]
At the moment there is a situation where `TPOTRegressor.predict(X)` and `TPOTRegressor.fitted_pipeline_.predict(X)` exhibit different behaviour - the former is able to make a prediction, whereas the latter fails with an error (if the testing dataset contains missing values). Intuitively, they should behave the same.

Or am I missing some important TPOT design consideration?

**Comment by rhiever** [comment]
We didn't intend for the `fitted_pipeline_` property to be interacted with in that way, although I can see why that's an issue. The alternative solution is to inject an Imputer at the beginning of every pipeline when they're internally generated, but we'll need to add extra logic that covers pipeline splits too.

cc @weixuanfu

**Comment by vruusmann** [comment]
I'm fetching the value of `fitted_pipeline_` in order to send it to the PMML conversion utility:
```Python
from sklearn2pmml import make_pmml_pipeline, sklearn2pmml

estimator = TPOTClassifier(...)
estimator.fit(X, y)
pmml_pipeline = make_pmml_pipeline(estimator.fitted_pipeline_)

sklearn2pmml(pmml_pipeline, "Pipeline.pmml")
```

Everything seems to work fine (ie. TPOT and (J)PMML predictions are 100% the same) when the dataset contains numeric features:
https://github.com/jpmml/jpmml-sklearn/blob/master/src/test/resources/extensions.py

I still need to figure out a solution for the following two problem areas:
- How to prepare categorical features? Ideally, I'd like to be able to pass string columns (ie. categorical features) directly to the `TPOTEstimator.fit(X, y)` method.
- How to prepare missing values (both continuous and categorical features)? I can see from the above comments that TPOT estimator classes perform value imputation internally. However, it would be necessary to expose the state of fitted imputers to end users somehow (so that the PMML conversion utility could write down the corresponding logic) - why not insert them into the first position of the `fitted_pipeline_` attribute?

**Comment by rhiever** [comment]
Good question. What do you do with regular sklearn pipelines? My first thoughts are:

* Use a LabelEncoder to convert string columns to corresponding numerical columns before passing them to the pipeline.

* Use a fixed imputation strategy (i.e., Imputer with the 'median' strategy) on datasets with missing data before passing them to the pipeline.

The above recommendations are in line with his sklearn works: sklearn assumes that the data is complete (i.e., no missingness) and numerically encoded. It leaves the handling of datasets with incomplete or string values to the user.

Maybe @amueller would have some useful comments here, as this issue goes beyond just TPOT; it applies to all sklearn pipelines.

**Comment by amueller** [comment]
For categorical we have CategoricalEncoder in master so you don't need LabelEncoder.

I don't understand what the question is. How to convert an sklearn pipeline to PMML?
It looks like ``make_pmml_pipeline`` does that, but ``estimator.fitted_pipeline_`` doesn't contain all the steps, and so converting it to pmml is not the same as converting ``estimator`` to PMML.

I think @vruusmann is saying that if TPOT put all steps into  ``estimator.fitted_pipeline_``  he could just use ``make_pmml_pipeline``, but since it doesn't, you'd need a custom PMML exporter for ``TPOTClassifier``.

**Comment by rhiever** [comment]
👍 on LabelEncoder!

wrt the imputer: I agree, and it's worth filing an issue to rework how TPOT handles missing data. I don't believe it's as simple as injecting an Imputer object at the beginning of every pipeline because TPOT can have "splits" in the pipeline that work on separate copies of the original dataset (as seen [here](https://github.com/rhiever/tpot/blob/master/images/tpot-pipeline-example.png)).

**Comment by vruusmann** [comment]
The TPOT project appears to contain a `tpot.builtins.OneHotEncoder` transformation. However, I couldn't find any information about by whom/where/how it is incorporated into the pipeline - am expecting that TPOT contains some "detect categorical columns"-logic, and does it all for end users.

Suppose my input dataset contains an integer column (eg. obtained by applying `LabelEncoder` to a string column) that needs to be "interpreted" as categorical not continuous. How can I order TPOT to apply this `tpot.builtins.OneHotEncoder` transformation to it? Or, more generally, how can I order TPOT to split the "mixed" original dataset into "clean" categorical and continuous sub-datasets for feature engineering work?

**Comment by rhiever** [comment]
`tpot.builtins.OneHotEncoder` was adopted from auto-sklearn some time ago. It's treated as just another preprocessing operator and is integrated into the [sparse config](https://github.com/rhiever/tpot/blob/5516d7ef501c39efb59c2dd1ec7ba100cbe540c4/tpot/config/classifier_sparse.py#L29). This OHE implementation considers all columns with <= 10 unique values to be categorical (that threshold value is a tunable parameter), and all other columns to be continuous. It will apply OHE only to categorical columns.

It also handles missing data and treats it as a another level in the categorical features.

**Comment by villasv** [comment]
@rhiever 
> TPOT's imputation step isn't stored in the internal pipelines, but it should apply imputation when you call any functions like fit, predict, score, etc.

Here's the code inside `TPOTBase(BaseEstimator).predict`

```
features = features.astype(np.float64)

if np.any(np.isnan(features)):
    self._imputed = True
    features = self._impute_values(features)
else:
    self._imputed = False
return self.fitted_pipeline_.predict(features)
```

So yeah, it will use the `self._fitted_imputer` when predicting because of the `self._impute_values` call.

But now here's the code inside `TPOTBase(BaseEstimator).score`:

```
score = SCORERS[self.scoring_function](
            self.fitted_pipeline_,
            testing_features.astype(np.float64),
            testing_target.astype(np.float64)
)
```

It simply forwards the `testing_features` without using the `self._fitted_imputer`, so unlike I expected, it won't do imputation automatically when using the `score` method.

Should I open a new issue? (And perhaps a PR...)
